src/xdp-utils.c

```diff
@@ -163,16 +164,19 @@ _xdp_parse_app_id_from_unit_name (const char *unit)
   /*
    * From https://systemd.io/DESKTOP_ENVIRONMENTS/ the format is one of:
    * app[-<launcher>]-<ApplicationID>-<RANDOM>.scope
+   * app[-<launcher>]-<ApplicationID>-<RANDOM>.slice
+   */
+  regex1 = g_regex_new ("^app-(?:[[:alnum:]]+\\-)?(.+?)(?:\\-[[:alnum:]]*)(?:\\.scope|\\.slice)$", 0, 0, &error);
+  g_assert (error == NULL);
+  /*
    * app[-<launcher>]-<ApplicationID>-autostart.service -> no longer true since systemd v248
    * app[-<launcher>]-<ApplicationID>[@<RANDOM>].service
-   * app[-<launcher>]-<ApplicationID>-<RANDOM>.slice
    */
-  regex = g_regex_new ("^app-(?:[[:alnum:]]+\\-)?(.+?\\..+?\\..+?)"
-                       "(?:[@\\-][[:alnum:]]*)?(?:-autostart)?"
-                       "(?:\\.scope|\\.service|\\.slice)$",
-                       0, 0, &error);
+  regex2 = g_regex_new ("^app-(?:[[:alnum:]]+\\-)?(.+?)(?:@[[:alnum:]]*|\\-autostart)?\\.service$", 0, 0, &error);
   g_assert (error == NULL);
-  if (g_regex_match (regex, unit, 0, &match))
+
+  if (g_regex_match (regex1, unit, 0, &match) ||
+      g_regex_match (regex2, unit, 0, &match))
```

Comment:
This leaks:

> Note that if match_info is not NULL then it is created even if the function returns FALSE, i.e. you must free it regardless if regular expression actually matched

The cleanest way I can think of is something like this:

```suggestion
  if (!g_regex_match (regex1, unit, 0, &match))
    g_clear_pointer (&match, g_match_info_unref);

  if (match == NULL && !g_regex_match (regex2, unit, 0, &match))
    g_clear_pointer (&match, g_match_info_unref);

  if (match != NULL)
```

Reply:
Good catch!